core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala

```diff
@@ -96,57 +97,55 @@ class ReplicaAlterLogDirsThread(name: String,
   }
 
   override def removePartitions(topicPartitions: Set[TopicPartition]): Map[TopicPartition, PartitionFetchState] = {
-    // Schedule assignment request to revert any queued request before cancelling
-    for {
-      topicPartition <- topicPartitions
-      partitionState <- partitionAssignmentRequestState(topicPartition)
-      if partitionState == QUEUED
-      partition = replicaMgr.getPartitionOrException(topicPartition)
-      topicId <- partition.topicId
-      directoryId <- partition.logDirectoryId()
-      topicIdPartition = new TopicIdPartition(topicId, topicPartition.partition())
-    } directoryEventHandler.handleAssignment(topicIdPartition, directoryId, () => ())
+    for (topicPartition <- topicPartitions) {
+      // Revert any reassignments for partitions that did not complete the future replica promotion
+      val PromotionState(reassignmentState, topicId, originalDir) = this.promotionState(topicPartition)
+      if (reassignmentState.inconsistentMetadata) {
+        directoryEventHandler.handleAssignment(new TopicIdPartition(topicId, topicPartition.partition()), originalDir, () => ())
+      }
+
+      this.promotionStates.remove(topicPartition)
+    }
 
     super.removePartitions(topicPartitions)
   }
 
+  private def promotionState(topicPartition: TopicPartition): PromotionState = promotionStates.get(topicPartition)
+
+  private def reassignmentState(topicPartition: TopicPartition): ReassignmentState = promotionState(topicPartition).reassignmentState
+
   // Visible for testing
-  private[server] def updatedAssignmentRequestState(topicPartition: TopicPartition)(state: ReplicaAlterLogDirsThread.DirectoryEventRequestState): Unit = {
-    assignmentRequestStates.put(topicPartition, state)
+  private[server] def updateReassignmentState(topicPartition: TopicPartition, state: ReassignmentState): Unit = {
+    promotionStates.put(topicPartition, promotionState(topicPartition).withAssignment(state))
   }
 
   private def maybePromoteFutureReplica(topicPartition: TopicPartition, partition: Partition) = {
     val topicId = partition.topicId
     if (topicId.isEmpty)
       throw new IllegalStateException(s"Topic ${topicPartition.topic()} does not have an ID.")
 
-    partitionAssignmentRequestState(topicPartition) match {
-      case None =>
+    reassignmentState(topicPartition) match {
+      case ReassignmentState.None =>
         // Schedule assignment request and don't promote the future replica yet until the controller has accepted the request.
         partition.runCallbackIfFutureReplicaCaughtUp(_ => {
```

Comment:
Should we log something in debug level or trace level, if `runCallbackIfFutureReplicaCaughtUp` return false, because of not caught-up?

Reply:
I think that would be a very noisy log, this method is called for every fetched batch. I'm not sure how it would be useful. What do you have in mind?

Reply:
What I'm afraid is the future log, somehow, it never get promoted because it cannot catch up with the local log, users might want to have clue from the log output. WDYT? If you think it's too noisy, we can leave it as is, and add it in the future if necessary.

Reply:
Yes, I think so. I'll add a debug log to `updateReassignmentState`, so we can log every state transition. In that situation, if we don't see the respective transition change that would indicate the future log has not caught up.

Reply:
Sounds good to me. Thanks.